scripts/cfos/validate.py

```diff
@@ -5,70 +5,256 @@
-                                                Column("APOE_value", [IsDtypeValidation(int)]),
-                                                Column("APOE_atDonorAge", [IsDtypeValidation(int)]),
-                                                Column("DataModality", [InListValidation(data_modality_options)]),
-                                                # file path columns must start with "gs://"
-                                                # TODO: check starts with not just for pattern
-                                                Column("summary_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("features_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("matrix_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("barcode_file_path", [MatchesPatternValidation("gs://")]),
-                                                Column("biosample_id_of_file", null_validation),
-                                                Column("library_id", null_validation),
-                                                Column("UMI_threshold", [IsDtypeValidation(int)]),
-                                                Column("follow_up_years", [IsDtypeValidation(int)]),
-                                                Column("has_disease", [InListValidation(has_disease_options)])
-                                                   #    Column("comment", [None])
-                                            ])
+null_validation = CustomElementValidation(lambda d: d is not np.nan, 'this field cannot be null')
 
+      
+def field_exists_in_df(field_id, df):
+    if field_id in list(df.keys()):
+        return True
+    return False
 
 
+# unclear if this is necessary
+def required_field(field_id, field_dict):
+   if field_id not in list(field_dict.keys()):
+      return False
+   field_id_keys = list(field_dict[field_id].keys())
 
+   if get_optional_field_key() not in field_id_keys:
+      return True
 
+   if field_attribute_is_true(attribute_to_check=get_optional_field_key(), field_id=field_id, field_dict=field_dict):
+      return False
 
+   return True
 
 
+def field_attribute_is_true(field_id, field_dict, attribute_to_check):
+   # handle null defaults to "true" on a field-by-field basis outside this function
+   if attribute_to_check not in list(field_dict[field_id].keys()):
+      return False
+
+   if field_dict[field_id][attribute_to_check] == "True":
+      return True
+   return False
+
+
+def field_attribute_value_exists(field_id, field_dict, attribute_to_check):
+   if attribute_to_check not in list(field_dict[field_id].keys()):
+      return False
+
+   attribute_value = str(field_dict[field_id][attribute_to_check]).strip()
+   if attribute_value is None or attribute_value == "":
```

Comment:
this can be simplified to `return attribute_value is not None and attribute_value != ""`

Reply:
TODO: Update to 'return attribute_value != ""'

Reply:
This should be updated now!